File: android/src/main/java/com/margelo/nitro/svga/HybridSvgaManager.kt
package com.margelo.nitro.svga

import android.util.Log
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.cancel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit

@DoNotStrip
@Keep
class HybridSvgaManager : HybridSvgaManagerSpec() {

  private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
  private val preloadGate = Semaphore(MAX_CONCURRENT_PRELOADS)
  private val context get() = NitroModules.applicationContext
    ?: throw IllegalStateException("Application context unavailable")
  private val sounds = SvgaSoundLibrary(context)

  init { SvgaMemoryCache.ensureInit(context) }

  override fun preload(urls: Array<String>, cacheKeys: Array<String>): Promise<Unit> {
    return Promise.async(scope) {
      coroutineScope {
        urls.mapIndexed { i, source ->
          val key = effectiveKey(source, cacheKeys, i)
          async { preloadGate.withPermit { preloadOne(source, key) } }
        }.awaitAll()
      }
      Unit
    }
  }

  override fun preloadDecoded(urls: Array<String>, cacheKeys: Array<String>): Promise<Unit> {
    return Promise.async(scope) {
      // Best-effort warmup — failures don't fail the whole batch (one bad
      // URL shouldn't kill 100 valid preloads), but we log them so they're
      // not silently invisible. Previously this swallowed every Throwable
      // with no signal at all.
      coroutineScope {
        urls.mapIndexed { i, source ->
          val key = effectiveKey(source, cacheKeys, i)
          async {
            preloadGate.withPermit {
              try {
                // loadEntity returns +1 ownership; we only want to warm
                // the cache (which holds its own +1), so release ours
                // immediately. Without this the entity bitmaps would
                // never recycle until the JS context tore down.
                SvgaSourceLoader.loadEntity(context, source, key).release()
              } catch (e: CancellationException) {
                throw e
              } catch (t: Throwable) {
                Log.w(TAG, "preloadDecoded failed for $source: ${t.message}")
              }
            }
          }
        }.awaitAll()
      }
      Unit
    }
  }

  override fun isCached(cacheKey: String): Boolean {
    val resolved = UrlValidator.resolve(cacheKey)
      ?: return SvgaDiskCache.isCached(context, cacheKey)
    return when (resolved.kind) {
      UrlValidator.Kind.REMOTE -> SvgaDiskCache.isCached(context, cacheKey)
      UrlValidator.Kind.LOCAL_FILE -> java.io.File(resolved.value).isFile
      UrlValidator.Kind.BUNDLED_ASSET -> assetExists(resolved.value)
    }
  }

  override fun getCachePath(cacheKey: String): String? {
    val resolved = UrlValidator.resolve(cacheKey)
      ?: return SvgaDiskCache.pathOrNull(context, cacheKey)
    return when (resolved.kind) {
      UrlValidator.Kind.REMOTE -> SvgaDiskCache.pathOrNull(context, cacheKey)
      UrlValidator.Kind.LOCAL_FILE -> java.io.File(resolved.value).takeIf { it.isFile }?.absolutePath
      UrlValidator.Kind.BUNDLED_ASSET -> if (assetExists(resolved.value)) resolved.value else null
    }
  }

  private fun assetExists(name: String): Boolean {
    return try {
      context.assets.open(name).use { true }
    } catch (_: java.io.IOException) {
      false
    }
  }

  override fun clearCache() {
    // Memory clear is cheap and synchronous so the next memory-cache read
    // sees a miss immediately. The on-disk delete walks the cache dirs and
    // can list+unlink hundreds of files on a busy device — offload to the
    // IO scope so it can't ANR the JS thread.
    SvgaMemoryCache.clear()
    scope.launch {
      try {
        SvgaDiskCache.clearSvga(context)
        SvgaDiskCache.clearSounds(context)
      } catch (t: Throwable) {
        Log.w(TAG, "clearCache disk eviction failed: ${t.message}")
      }
    }
  }

  override fun getCacheSize(): Promise<Double> {
    return Promise.async(scope) { SvgaDiskCache.totalSvgaBytes(context).toDouble() }
  }

  override fun getCacheCount(): Promise<Double> {
    return Promise.async(scope) { SvgaDiskCache.totalSvgaCount(context).toDouble() }
  }

  override fun setCacheLimit(bytes: Double) {
    SvgaDiskCache.setMaxBytes(clampToLong(bytes))
  }

  override fun setMemoryLimit(bytes: Double) {
    SvgaMemoryCache.setMaxBytes(clampToLong(bytes))
  }

  override fun setMaxAgeMs(ms: Double) {
    val safe = clampToLong(ms)
    // Mirror to both layers so a hit at either level honours the TTL.
    SvgaDiskCache.setMaxAgeMs(safe)
    SvgaMemoryCache.setMaxAgeMs(safe)
  }

  /// Defensive clamp against `NaN`/`Infinity`/negative `Double`s. The JS
  /// `assertNonNegativeFinite` guard catches misuse at the package boundary,
  /// but a non-Nitro caller reaching this spec from another module would
  /// otherwise overflow into `Long.MIN_VALUE` (NaN.toLong() returns 0; +Inf
  /// returns Long.MAX_VALUE on JVM, but defensive code reads cleaner).
  private fun clampToLong(value: Double): Long {
    if (value.isNaN() || value <= 0.0) return 0L
    if (value >= Long.MAX_VALUE.toDouble()) return Long.MAX_VALUE
    return value.toLong()
  }

  override fun evictExpired(): Promise<Double> {
    return Promise.async(scope) { SvgaDiskCache.evictExpired(context).toDouble() }
  }

  override fun loadSound(key: String, url: String): Promise<Unit> {
    return Promise.async(scope) {
      val file = SvgaSourceLoader.loadSoundBytes(context, url)
      sounds.load(key, file)
    }
  }

  override fun playSound(key: String, volume: Double) {
    sounds.play(key, volume.toFloat())
  }

  override fun stopSound(key: String) { sounds.stop(key) }
  override fun stopAllSounds() { sounds.stopAll() }
  override fun unloadSound(key: String) { sounds.unload(key) }

  override fun dispose() {
    scope.cancel()
    sounds.release()
  }

  /// Pick the effective cache key from `cacheKeys[i]`, falling back to the
  /// URL when the JS layer didn't supply one (or the array is short — defend
  /// against malformed callers).
  private fun effectiveKey(url: String, cacheKeys: Array<String>, i: Int): String {
    val raw = if (i < cacheKeys.size) cacheKeys[i] else ""
    return if (raw.isNotEmpty()) raw else url
  }

  private suspend fun preloadOne(source: String, cacheKey: String) {
    val resolved = UrlValidator.resolve(source) ?: return
    if (resolved.kind != UrlValidator.Kind.REMOTE) return
    SvgaSourceLoader.preloadRemote(context, resolved.value, cacheKey)
  }

  companion object {
    private const val MAX_CONCURRENT_PRELOADS = 4
    private const val TAG = "SvgaManager"
  }
}

File: ios/HybridSvgaManager.swift
import Foundation
import NitroModules

final class HybridSvgaManager: HybridSvgaManagerSpec {

    private static let MAX_CONCURRENT_PRELOADS = 4
    private let sounds = SvgaSoundLibrary()

    func preload(urls: [String], cacheKeys: [String]) throws -> Promise<Void> {
        // The manager is owned by Nitro for the JS context's lifetime; if we
        // captured `self` weakly and Nitro reclaimed the manager mid-preload,
        // the promise would resolve with `()` instead of failing — JS-side
        // `await SvgaCache.preload(urls)` would then return a false success.
        // Capture strongly and rely on `cancel()` in `dispose()` to break
        // outstanding work cleanly.
        return Promise.async {
            try await self.preloadAll(urls, cacheKeys: cacheKeys)
        }
    }

    func preloadDecoded(urls: [String], cacheKeys: [String]) throws -> Promise<Void> {
        let limit = Self.MAX_CONCURRENT_PRELOADS
        return Promise.async {
            // Collect the first error and propagate it once the group drains
            // (so a single bad URL surfaces up to JS as a real failure
            // instead of silently swallowing every result). Other URLs are
            // best-effort and don't get re-thrown.
            try await withThrowingTaskGroup(of: Result<Void, Error>.self) { group in
                var inFlight = 0
                let pairs = Self.zipKeys(urls: urls, cacheKeys: cacheKeys)
                var iterator = pairs.makeIterator()
                func enqueueNext() {
                    guard let (source, key) = iterator.next() else { return }
                    inFlight += 1
                    group.addTask {
                        await withCheckedContinuation { (cont: CheckedContinuation<Result<Void, Error>, Never>) in
                            SvgaSourceLoader.loadEntity(source, cacheKey: key) { result in
                                switch result {
                                case .success: cont.resume(returning: .success(()))
                                case .failure(let err): cont.resume(returning: .failure(err))
                                }
                            }
                        }
                    }
                }
                for _ in 0..<min(limit, pairs.count) { enqueueNext() }
                var firstError: Error?
                while inFlight > 0 {
                    if let result = try await group.next() {
                        if case .failure(let err) = result, firstError == nil {
                            firstError = err
                        }
                    }
                    inFlight -= 1
                    enqueueNext()
                }
                if let firstError = firstError { throw firstError }
            }
        }
    }

    func isCached(cacheKey: String) throws -> Bool {
        // For local/bundled paths the cacheKey is the filesystem path itself,
        // not a remote-cache identity — keep the existing behaviour for those.
        guard let resolved = UrlValidator.resolve(cacheKey) else {
            return SvgaDiskCache.isCached(cacheKey)
        }
        switch resolved.kind {
        case .remote:
            return SvgaDiskCache.isCached(cacheKey)
        case .localFile:
            return FileManager.default.fileExists(atPath: resolved.value)
        case .bundledAsset:
            return Self.assetExists(resolved.value)
        }
    }

    func getCachePath(cacheKey: String) throws -> String? {
        guard let resolved = UrlValidator.resolve(cacheKey) else {
            return SvgaDiskCache.pathOrNil(cacheKey)
        }
        switch resolved.kind {
        case .remote:
            return SvgaDiskCache.pathOrNil(cacheKey)
        case .localFile:
            return FileManager.default.fileExists(atPath: resolved.value) ? resolved.value : nil
        case .bundledAsset:
            return Self.assetExists(resolved.value) ? resolved.value : nil
        }
    }

    private static func assetExists(_ name: String) -> Bool {
        return SvgaSourceLoader.bundleURL(for: name) != nil
    }

    func clearCache() throws {
        try SvgaDiskCache.clearSvga()
        try SvgaDiskCache.clearSounds()
        SvgaMemoryCache.shared.clear()
    }

    func getCacheSize() throws -> Promise<Double> {
        return Promise.async {
            return Double(SvgaDiskCache.totalSvgaBytes())
        }
    }

    func getCacheCount() throws -> Promise<Double> {
        return Promise.async {
            return Double(SvgaDiskCache.totalSvgaCount())
        }
    }

    func setCacheLimit(bytes: Double) throws {
        SvgaDiskCache.setMaxBytes(Self.clampToInt64(bytes))
    }

    func setMemoryLimit(bytes: Double) throws {
        SvgaMemoryCache.shared.setMaxBytes(Self.clampToInt(bytes))
    }

    /// Defensive clamp against non-finite or out-of-range `Double`s. The JS
    /// `assertNonNegativeFinite` guard catches misuse at the package
    /// boundary, but a non-Nitro caller reaching the spec from another
    /// module could otherwise trap on `Int64(.infinity)` / `Int(.nan)`.
    private static func clampToInt64(_ value: Double) -> Int64 {
        if !value.isFinite || value <= 0 { return 0 }
        if value >= Double(Int64.max) { return Int64.max }
        return Int64(value)
    }
    private static func clampToInt(_ value: Double) -> Int {
        if !value.isFinite || value <= 0 { return 0 }
        if value >= Double(Int.max) { return Int.max }
        return Int(value)
    }

    func setMaxAgeMs(ms: Double) throws {
        // Mirror to both layers so a hit at either level honours the TTL.
        let clamped = Self.clampToInt64(ms)
        SvgaDiskCache.setMaxAgeMs(clamped)
        SvgaMemoryCache.shared.setMaxAgeMs(clamped)
    }

    func evictExpired() throws -> Promise<Double> {
        return Promise.async {
            // Bridge the callback-style API into Swift Concurrency so this
            // Task suspends (rather than parking a cooperative-pool worker)
            // while the disk walk runs on writeQueue.
            return await withCheckedContinuation { (cont: CheckedContinuation<Double, Never>) in
                SvgaDiskCache.evictExpired { count in
                    cont.resume(returning: Double(count))
                }
            }
        }
    }

    func loadSound(key: String, url: String) throws -> Promise<Void> {
        return Promise.async {
            let url = try await self.fetchSoundFile(key: key, url: url)
            try self.sounds.load(key: key, url: url)
        }
    }

    func playSound(key: String, volume: Double) throws {
        sounds.play(key: key, volume: Float(volume))
    }

    func stopSound(key: String) throws { sounds.stop(key: key) }
    func stopAllSounds() throws { sounds.stopAll() }
    func unloadSound(key: String) throws { sounds.unload(key: key) }

    func dispose() { sounds.release() }
    deinit { sounds.release() }

    private func preloadAll(_ urls: [String], cacheKeys: [String]) async throws {
        let limit = Self.MAX_CONCURRENT_PRELOADS
        let pairs = Self.zipKeys(urls: urls, cacheKeys: cacheKeys)
        try await withThrowingTaskGroup(of: Void.self) { group in
            var iterator = pairs.makeIterator()
            var inFlight = 0
            func enqueue() throws {
                guard let (source, key) = iterator.next() else { return }
                inFlight += 1
                group.addTask { try await self.preloadOne(source, cacheKey: key) }
            }
            for _ in 0..<min(limit, pairs.count) { try enqueue() }
            while inFlight > 0 {
                try await group.next()
                inFlight -= 1
                try enqueue()
            }
        }
    }

    private func preloadOne(_ source: String, cacheKey: String) async throws {
        guard let resolved = UrlValidator.resolve(source) else { return }
        if resolved.kind != .remote { return }
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            SvgaSourceLoader.preloadRemote(resolved.value, cacheKey: cacheKey) { result in
                switch result {
                case .success: cont.resume(returning: ())
                case .failure(let err): cont.resume(throwing: err)
                }
            }
        }
    }

    private func fetchSoundFile(key: String, url: String) async throws -> URL {
        return try await withCheckedThrowingContinuation { (cont: CheckedContinuation<URL, Error>) in
            SvgaSourceLoader.loadSoundFile(key: key, url: url) { result in
                switch result {
                case .success(let url): cont.resume(returning: url)
                case .failure(let err): cont.resume(throwing: err)
                }
            }
        }
    }

    /// Pair urls with cache keys, falling back to the URL when the
    /// corresponding cacheKey is empty or missing. The JS layer normalises so
    /// in practice arrays are always aligned and non-empty, but defend
    /// against length mismatch to avoid an out-of-bounds crash on a malformed
    /// caller.
    private static func zipKeys(urls: [String], cacheKeys: [String]) -> [(String, String)] {
        var out: [(String, String)] = []
        out.reserveCapacity(urls.count)
        for i in 0..<urls.count {
            let url = urls[i]
            let key = (i < cacheKeys.count && !cacheKeys[i].isEmpty) ? cacheKeys[i] : url
            out.append((url, key))
        }
        return out
    }
}
